### DISCUSSION TITLE
How can I group nodes and how can I insert nodes into a parent node?

### DISCUSSION THREAD

**Discussion by dkrusenstrahle** [ueberdosis/tiptap#1270] [General Questions]
Hello,

Using Tiptap 2.0. Is it possible to add a child node into a parent node?
I want to make a list/group of nodes (not a bulletlist).

In other words, how can I group nodes and how can I insert nodes into a parent node?

**Comment by philippkuehn** [comment]
This is a complex topic 😅 Have you tried to use the [insertContent](https://www.tiptap.dev/api/commands/insert-content) command?

**Reply by dkrusenstrahle** [reply]
Yes very complex. I know how to insert content. The problem is how to insert at the right place meaning below the current node but within their parent. 

How can I get the current position and perhaps the parent?

**Reply by philippkuehn** [reply]
You should probably look into resolved positions:

```js
editor.state.selection.$anchor
editor.state.selection.$from
editor.state.selection.$to
```

Within a `ResolvedPos` there are information about parent nodes and more.

You can get a resolved position for any position in your document.
```js
editor.state.doc.resolve(25)
```

**Reply by dkrusenstrahle** [reply]
Interesting! How can I use this data?
For example, how can I insert a new node below the current/selected node based upon the information I get from the above code?

**Reply by philippkuehn** [reply]
Since the latest release there is a new command called `insertContentAt`. This can be very helpful because with that you can insert nodes at a specific position.

```js
// insert single node
editor.commands.insertContentAt({ from: 10, to: 10 }, {
  type: 'paragraph',
  content: [
    {
      type: 'text',
      text: 'A paragraph',
    },
  ],
})

// insert multiple paragraphs
editor.commands.insertContentAt({ from: 10, to: 10 }, [
  {
    type: 'paragraph',
    content: [
      {
        type: 'text',
        text: 'First paragraph',
      },
    ],
  },
  {
    type: 'paragraph',
    content: [
      {
        type: 'text',
        text: 'Second paragraph',
      },
    ],
  },
])
```

with that in mind you can do something like that:

```js
editor
  .chain()
  .focus()
  .command(({ tr, commands }) => {
    // get position after the current parent node
    const position = tr.selection.$cursor.after()

    return commands.insertContentAt({ from: position, to: position }, {
      type: 'text',
      text: 'yay',
    })
  })
  .run()
```

**Reply by dkrusenstrahle** [reply]
Wow great! I just updated to 2.0.0-beta.44 but I can however not find the new "insertContentAt". Is there a newer version of the core than 2.0.0-beta.44? I have updated the other packages too (that I am using).

**Reply by philippkuehn** [reply]
Hmm, it is released in 2.0.0-beta.44 🤔

**Reply by dkrusenstrahle** [reply]
Strange! I got an error when making a new line in the editor after updating too (just filed an issued about it).

**Reply by dkrusenstrahle** [reply]
Made a clean install of the entire project and now I have access to insertContentAt.
Thank you!

**Reply by dkrusenstrahle** [reply]
I tried the following but there are no $cursor only $anchor. And when I try it I get this message.

`const position = tr.selection.$anchor.after();`

`ERROR RangeError: There is no position after the top-level node`

**Reply by philippkuehn** [reply]
I think at this point I can’t help you anymore without seeing any code. Maybe you could provide a simple codesandbox for it.

**Reply by dkrusenstrahle** [reply]
@philippkuehn I got this working. It replaces the current node and if I add + 1 to the to and from variables it replaces the node after which is not what I am looking for.

Is there a way to insert "between" the current node and the one after (keeping both and just inserting the new one)?

```
const from = this.props.editor.state.selection.from;
const to = this.props.editor.state.selection.to;
return this.props.editor.commands.insertContentAt({ from: from, to: to }, {
   type: 'text',
   text: 'yay',
})
```

**Reply by philippkuehn** [reply]
If you insert content at the end position of a node, this should not happen. But it's really hard to tell without a codesandbox or something.
